pkg/kubelet/cm/cgroup_manager_linux.go

```diff
@@ -574,176 +574,59 @@ func CpuWeightToCpuShares(cpuWeight uint64) uint64 {
-	cgroupResourcePath, found := cgroupPaths[string(resource)]
-	if !found {
-		return nil, fmt.Errorf("failed to build %v cgroup fs path for cgroup %v", resource, name)
+	cgroupResources := cgroups.Resources
+	if cgroupResources == nil {
+		return nil, fmt.Errorf("cgroup resources are nil for cgroup %v", name)
 	}
-	switch resource {
-	case v1.ResourceCPU:
-		return getCgroupCpuConfig(cgroupResourcePath)
-	case v1.ResourceMemory:
-		return getCgroupMemoryConfig(cgroupResourcePath)
-	}
-	return nil, fmt.Errorf("unsupported resource %v for cgroup %v", resource, name)
-}
 
-func setCgroupv1CpuConfig(cgroupPath string, resourceConfig *ResourceConfig) error {
-	var cpuQuotaStr, cpuPeriodStr, cpuSharesStr string
-	if resourceConfig.CPUQuota != nil {
-		cpuQuotaStr = strconv.FormatInt(*resourceConfig.CPUQuota, 10)
-		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.cfs_quota_us"), []byte(cpuQuotaStr), 0700); err != nil {
-			return fmt.Errorf("failed to write %v to %v: %v", cpuQuotaStr, cgroupPath, err)
-		}
-	}
-	if resourceConfig.CPUPeriod != nil {
-		cpuPeriodStr = strconv.FormatUint(*resourceConfig.CPUPeriod, 10)
-		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.cfs_period_us"), []byte(cpuPeriodStr), 0700); err != nil {
-			return fmt.Errorf("failed to write %v to %v: %v", cpuPeriodStr, cgroupPath, err)
-		}
-	}
-	if resourceConfig.CPUShares != nil {
-		cpuSharesStr = strconv.FormatUint(*resourceConfig.CPUShares, 10)
-		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.shares"), []byte(cpuSharesStr), 0700); err != nil {
-			return fmt.Errorf("failed to write %v to %v: %v", cpuSharesStr, cgroupPath, err)
-		}
-	}
-	return nil
+	return libcontainerCgroupResourcesToResourceConfig(*cgroupResources), nil
 }
 
-func setCgroupv2CpuConfig(cgroupPath string, resourceConfig *ResourceConfig) error {
-	if resourceConfig.CPUQuota != nil {
-		if resourceConfig.CPUPeriod == nil {
-			return fmt.Errorf("CpuPeriod must be specified in order to set CpuLimit")
-		}
-		cpuLimitStr := Cgroup2MaxCpuLimit
-		if *resourceConfig.CPUQuota > -1 {
-			cpuLimitStr = strconv.FormatInt(*resourceConfig.CPUQuota, 10)
-		}
-		cpuPeriodStr := strconv.FormatUint(*resourceConfig.CPUPeriod, 10)
-		cpuMaxStr := fmt.Sprintf("%s %s", cpuLimitStr, cpuPeriodStr)
-		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.max"), []byte(cpuMaxStr), 0700); err != nil {
-			return fmt.Errorf("failed to write %v to %v: %v", cpuMaxStr, cgroupPath, err)
-		}
-	}
-	if resourceConfig.CPUShares != nil {
-		cpuWeight := CpuSharesToCpuWeight(*resourceConfig.CPUShares)
-		cpuWeightStr := strconv.FormatUint(cpuWeight, 10)
-		if err := os.WriteFile(filepath.Join(cgroupPath, "cpu.weight"), []byte(cpuWeightStr), 0700); err != nil {
-			return fmt.Errorf("failed to write %v to %v: %v", cpuWeightStr, cgroupPath, err)
-		}
+// Set resource config for the specified resource type on the cgroup
+func (m *cgroupManagerImpl) SetCgroupConfig(name CgroupName, resourceConfig *ResourceConfig) error {
```

Comment:
I think that by getting rid of the resourceName parameter, we should expect this to be called once with the full set of resources, not iteratively as with the current implementation. I mention the issue with the CPU logic changing in some other comments. Another difference that stands out is that the device properties are always set on every call under the systemd implementation: https://github.com/opencontainers/runc/blob/v1.1.12/libcontainer/cgroups/systemd/common.go#L185-L190